### PR TITLE
request reset password and change password ui

### LINKED ISSUE #1034: User forgot password
- [x] IT + Controller + Repository
- [x] APIs
- [ ] Frontend 

integration test:

1. userA request to reset password 
userA changes is password 
userA login with his old password fails
userA login with his new password success

### PR DESCRIPTION
**Pull request by rfontanarosa**
close #1034

Schermata 2022-02-01 alle 09 36 05
Schermata 2022-02-01 alle 09 39 56
Schermata 2022-02-01 alle 10 36 42

### CODE REVIEW

File: server/api/auth/changePassword.ts [hunk 1]
```diff
@@ -12,18 +12,22 @@ export const AuthChangePassword = {
       try {
         const { email, password, password2, uuid } = req.body
 
-        if (Objects.isEmpty(password?.trim()) || Objects.isEmpty(password2?.trim())) Requests.send400(res, 'login.noEmptyPassword')
-        if (password?.trim() !== password2?.trim()) Requests.send400(res, 'login.noMatchPasswords')
-        if (!validPassword(password)) Requests.send400(res, 'login.passwordError')
+        if (Objects.isEmpty(password?.trim()) || Objects.isEmpty(password2?.trim()))
+          return Requests.send400(res, 'login.noEmptyPassword')
+        if (password?.trim() !== password2?.trim()) return Requests.send400(res, 'login.noMatchPasswords')
```
Comment (sorja) [inline server/api/auth/changePassword.ts:17]: backend handle only password, no password2

File: client/store/login/actions/changePassword.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import axios from 'axios'
+import { ApiEndPoint } from '@common/api/endpoint'
+
+export const changePassword = createAsyncThunk(
+  'login/post/changePassword',
+  async ({ email, password, password2, resetPasswordUuid }: any) => {
+    const { data } = await axios.post(ApiEndPoint.Auth.changePassword(), {
+      email,
+      password,
+      password2,
```
Comment (sorja) [inline client/store/login/actions/changePassword.ts:12]: backend handle only password, no password2

File: client/pages/Login/ResetPassword/ResetPassword.tsx [hunk 3]
```diff
@@ -1,37 +1,115 @@
...
+  const onChangePassword = async () => {
+    const fieldErrors = LoginValidator.localValidate(email, password, password2)
+    setErrors(fieldErrors)
+
+    if (!fieldErrors.isError) {
+      try {
+        await dispatch(LoginActions.changePassword({ email, password, resetPasswordUuid })).unwrap()
+        history.push(BasePaths.Root())
+      } catch (rejectedValueOrSerializedError) {
+        console.log(rejectedValueOrSerializedError)
```
Comment (minotogna) [inline client/pages/Login/ResetPassword/ResetPassword.tsx:50]: no console log

File: client/pages/Login/ResetPassword/ResetPassword.tsx [hunk 4]
```diff
@@ -1,37 +1,115 @@
...
+  const onResetPassword = async () => {
+    const fieldErrors = LoginValidator.resetPasswordValidate(email)
+    setErrors(fieldErrors)
+
+    if (!fieldErrors.isError) {
+      try {
+        await dispatch(LoginActions.createResetPassword(email)).unwrap()
+        history.push(BasePaths.Root())
+      } catch (rejectedValueOrSerializedError) {
+        console.log(rejectedValueOrSerializedError)
```
Comment (minotogna) [inline client/pages/Login/ResetPassword/ResetPassword.tsx:36]: no console.log

File: i18n/resources/en/login.js [hunk 5]
```diff
@@ -11,6 +11,7 @@ module.exports = {
   signInFRA: 'Sign in with FRA',
 
   resetPassword: 'Reset password',
+  changePassword: 'Change password',
```
Comment (minotogna) [inline i18n/resources/en/login.js:14]: add to githib notes missing translations

**Review by minotogna — CHANGES_REQUESTED** [review]

File: i18n/resources/en/login.js (hunk 5 above)
Comment (rfontanarosa) [reply i18n/resources/en/login.js:14]: changePassword is not missing, resetPassword is and it's already in the missing translations list

File: client/pages/Login/ResetPassword/ResetPassword.tsx [hunk 6]
```diff
@@ -1,37 +1,115 @@
...
+  const onResetPassword = async () => {
+    const fieldErrors = LoginValidator.resetPasswordValidate(email)
+    setErrors(fieldErrors)
+
+    if (!fieldErrors.isError) {
+      try {
+        await dispatch(LoginActions.createResetPassword(email)).unwrap()
+        history.push(BasePaths.Root())
+      } catch (rejectedValueOrSerializedError) {
+        // console.log(rejectedValueOrSerializedError)
```
Comment (sorja) [inline client/pages/Login/ResetPassword/ResetPassword.tsx:36]: Can we omit empty catch?
Or will the #1100 handle this?

**Review by sorja — COMMENTED** [review]
👍

File: client/pages/Login/ResetPassword/ResetPassword.tsx [hunk 7]
```diff
@@ -1,37 +1,115 @@
...
-  const onChangePassword = () => {
-    console.log('not implemented')
+  const onResetPassword = async () => {
+    const fieldErrors = LoginValidator.resetPasswordValidate(email)
+    setErrors(fieldErrors)
+
+    if (!fieldErrors.isError) {
+      try {
+        await dispatch(LoginActions.createResetPassword(email)).unwrap()
+        history.push(BasePaths.Root())
```
Comment (minotogna) [inline client/pages/Login/ResetPassword/ResetPassword.tsx:34]: do not use unwrap. pass history to action an d remove try catch

**Review by minotogna — CHANGES_REQUESTED** [review]

File: client/store/login/actions/changePassword.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import { createAsyncThunk } from '@reduxjs/toolkit'
+
+import axios from 'axios'
+import { ApiEndPoint } from '@common/api/endpoint'
+import { BasePaths } from '@client/basePaths'
+
+export const changePassword = createAsyncThunk(
+  'login/post/changePassword',
+  async ({ email, password, resetPasswordUuid, history }: any) => {
```
Comment (minotogna) [inline client/store/login/actions/changePassword.ts:9]: do not use any

File: client/store/login/actions/createResetPassword.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import { createAsyncThunk } from '@reduxjs/toolkit'
+
+import axios from 'axios'
+import { ApiEndPoint } from '@common/api/endpoint'
+import { BasePaths } from '@client/basePaths'
+
+export const createResetPassword: any = createAsyncThunk(
+  'login/post/createResetPassword',
+  async (email: string, history: any) => {
```
Comment (minotogna) [inline client/store/login/actions/createResetPassword.ts:9]: do not use any

**Review by minotogna — CHANGES_REQUESTED** [review]

File: client/store/login/actions/changePassword.ts [hunk 10]
```diff
@@ -1,12 +1,23 @@
...
+    email,
+    password,
+    resetPasswordUuid,
+    history,
+  }: {
+    email: string
+    password: string
+    resetPasswordUuid: string
+    history: H.History<H.LocationState>
+  }) => {
```
Comment (sorja) [inline client/store/login/actions/changePassword.ts:20]: Types should be defined in `createAsyncThunk<ret, params, ...>`

**Review by minotogna — APPROVED** [review]

**Pull request merged**